_posts/2018-04-10-personalized-search.md

```diff
@@ -0,0 +1,301 @@
+       String adid = doc.get("id");
+       return new ScoredAd(luceneDocumentId, adid, score);
+    }
+ 
+Next we use the recommendation api to get a personalization score for each document. This is a http-request, and is therefore also a high cost part, latency vise, of the plugin. 
+Our api takes the following parameters:
+- UserId
+- RecommenderId
+- List of AdIds (SOLR document ids)
+- MaxRows
+
+And returns a list of adIds with recommendation score. ScoredAd now contains both the SOLR- and the recommendation-score. These are further used for sorting the search results: ```total score = normalized solr score + (normalized recommendation score * recommendation score weight)```
+
+    final Map<String, ScoredAd> recommendedItems = getRecommendedItems(userId, recommenderId, reRankNum, results);
+    
+    if (recommendedItems.size() > 0) {
+       results.forEach(result -> {
+           final float normalizedSolrScore = normalizedSolrScore(result.getScore(), initialSearchResult.maxScore());
+           final float personalizationScore = personalizationScore(personalizationScoreWeight, recommendedItems, result.getAdId());
+           result.setScore(normalizedSolrScore + personalizationScore);
+       });
+       results.sort((r1, r2) -> Float.compare(r2.getScore(), r1.getScore()));
+    }
+
+Finally we are preparing the search response with a subset of the search results, set by offset and rows.
+
+    final List<ScoredAd> searchResult = results.stream()
+           .skip(offset)
+           .limit(rows)
+           .collect(toList());
+    try {
+       addToResponse(rb, reRankNum, searchResult, initialSearchResult);
+    } catch (IOException e) {
+       LOG.error("Could not add recommended items to search result: " + e);
+    }
+    
+    private static void addToResponse(ResponseBuilder rb, int reRankNum, List<ScoredAd> resultsToReturn, DocList docList) throws IOException {
+       DocListAndSet docListAndSet = buildDocListAndSet(reRankNum, resultsToReturn, docList);
+       rb.setResults(docListAndSet);
+    
+       BasicResultContext resultContext = new BasicResultContext(rb);
+       SolrQueryResponse rsp = rb.rsp;
+       rsp.getValues().removeAll("response");
+       rsp.addResponse(resultContext);
+    }
+
+
+## Response times
+We will also look at the response times:
+
+Response times for 95th percentile. Default search is searches without any custom plugins, distribution search is the actual search we are trying to replace with personalization search. The actual timings in the 95th percentile doubles from 225ms to 450ms. While the basic search takes around 80ms.
+
+<figure>
+    <img class="center-block" src="/images/2018-04-10-personalized-search/95th_percentile_latency.png" alt="alt" title="95th percentile latency" />
+    <figcaption style="text-align:center; font-style:italic;">95th percentile latency</figcaption>
```

Comment:
in the final view this looks like a heading of the next paragraph.  tune to make it look like the caption of the picture....